packaging_automation/citus_package.py

```diff
@@ -0,0 +1,264 @@
+import argparse
+import glob
+import os
+import subprocess
+from enum import Enum
+from typing import List
+from typing import Tuple
+import gnupg
+import docker
+
+from parameters_validation import non_blank, non_empty, validate_parameters
+
+from .common_tool_methods import (run_with_output, PackageType, transform_key_into_base64_str,
+                                  get_gpg_fingerprints_by_name, str_array_to_str)
+from .packaging_warning_handler import validate_output
+
+GPG_KEY_NAME = "packaging@citusdata.com"
+
+supported_platforms = {
+    "debian": ["buster", "stretch", "jessie", "wheezy"],
+    "el": ["8", "7", "6"],
+    "ol": ["7", "8"],
+    "ubuntu": ["focal", "bionic", "xenial", "trusty"]
+}
+
+
+def platform_names() -> List[str]:
+    platforms = []
+    for platform_os in supported_platforms:
+        for platform_release in supported_platforms[platform_os]:
+            platforms.append(f"{platform_os}/{platform_release}")
+    platforms.append("pgxn")
+    return platforms
+
+
+docker_image_names = {
+    "debian": "debian",
+    "el": "centos",
+    "ol": "oraclelinux",
+    "ubuntu": "ubuntu",
+    "pgxn": "pgxn"
+}
+
+
+def get_package_type_by_docker_image_name(docker_image_name: str) -> PackageType:
+    return PackageType.deb if docker_image_name.startswith(("ubuntu", "debian")) else PackageType.rpm
+
+
+class BuildType(Enum):
+    release = 1
+    nightly = 2
+
+
+class PostgresVersionDockerImageType:
+    multiple = 1,
+    single = 2
+
+
+platform_postgres_version_source = {
+    "el": PostgresVersionDockerImageType.multiple,
+    "ol": PostgresVersionDockerImageType.multiple,
+    "debian": PostgresVersionDockerImageType.single,
+    "ubuntu": PostgresVersionDockerImageType.single,
+    "pgxn": PostgresVersionDockerImageType.single
+}
+
+PKGVARS_FILE = "pkgvars"
+SINGLE_DOCKER_POSTGRES_PREFIX = "all"
+PACKAGES_DIR_NAME = "packages"
+
+
+def decode_os_and_release(platform_name: str) -> Tuple[str, str]:
+    parts = platform_name.split("/")
+
+    if len(parts) == 0 or len(parts) > 2 or (len(parts) == 1 and parts[0] != "pgxn"):
+        raise ValueError("Platforms should have two parts divided by '/' or should be 'pgxn' ")
+    elif len(parts) == 1 and parts[0] == "pgxn":
+        os_name = "pgxn"
+        os_release = ""
+    else:
+        os_name = parts[0]
+        os_release = parts[1]
+        if os_name not in supported_platforms:
+            raise ValueError(
+                f"{os_name} is not among supported operating systems. Supported operating systems are as below:\n "
+                f"{','.join(supported_platforms.keys())}")
+        if os_release not in supported_platforms[os_name]:
+            raise ValueError(f"{os_release} is not among supported releases for {os_name}."
+                             f"Supported releases are as below:\n {','.join(supported_platforms[os_name])}")
+    return os_name, os_release
```

Comment:
Instead of having all these logic in here, we can let all the platform_name variables store (os_name, release_name) tuples that will make it unnecessary to decode anything.

It does not really make sense to port the Perl implementation of this function here because python supports tuples and it can make it way easier to read and understand the logic here.

FYI, this is an answer to your question https://github.com/citusdata/tools/pull/119#discussion_r661135483